### ISSUE TITLE
A recap so far

### ISSUE THREAD

**Issue by alnitak** [alnitak/flutter_soloud_web#2]
What I have reached so far despite my complete lack of javascript/web knowledge:

- compiled C/C++ to one .wasm and one .js files
- compiled what in `flutter_soloud` is the isolate sources into a .js (using `dart compile js ...`)
- conditional imports of the `original solound_controller.dart` and `binding_player.dart`
- a class that manages an Isolate or a web Worker depending on the platform
- a class to import a .js file (the compiled `audio_isolate.dart`) and put it in the HTML DOM

What is working:
- calls to WASM functions bound in the compiled .js. The player can play
- compile `audio_isolate.dart` to js and use it as a web Worker
- send and receive messages from/to web worker

What is not working (as far as I understand):
- send/receive classes (like `SoundHash` and `AudioSource`)
- send/receive enums (like `WaveForm` and `PlayerErrors`). A workaround should be to use their `index` or `name`
- the SoLoud library loaded in the main Flutter isolate, is not accessible from the web worker
- [web audio is not supported inside a web worker](https://stackoverflow.com/questions/67949831/is-there-a-way-to-use-audiocontext-in-a-web-worker)

The latter is the main blocker now.
A solution could be the use of `AudioWorklets` and [Emscripten has added support for them](https://github.com/emscripten-core/emscripten/blob/main/ChangeLog.md#3132---021723). But I think it's used as an audio callback to manage audio data when received and not to interface WebAudio inside a worker.
Miniaudio supports `AudioWorkletNode` since a while by adding `-DMA_ENABLE_AUDIO_WORKLETS -sAUDIO_WORKLET=1 -sWASM_WORKERS=1 -sASYNCIFY` flags to the emscripten compiler. It generates 2 more .js files that should be loaded inside the web worker(?). I tried but the browser gives me errors.

`audio_isolate.dart` has to:
1. load audio files (this is time-consuming and should be run into an Isolate/Worker)
2. manage the current active sounds
3. monitor their state every 10ms

I don't know if 3. is really possible: we use SoLoud bindings in the main isolate and in the audio isolate. From both threads, we have access to the same instance of the C/C++ library. But on the web, we should import the js library from both the main isolate and from the web worker. 
Because of this, I don't know if, for example, we do an initEngine/loadFile/play within the main isolate, and then from the web worker the `getIsValidVoiceHandle()` will be able to do its job.

**Comment by alnitak** [comment]
In the last few days, I have managed to test point 3. Indeed, loading the .js both in the main thread and in the Web Worker creates 2 different instances and calling `getIsValidVoiceHandle()` does not obtain the desired result.

In this scenario, I don't know how to proceed. Once the .js has been loaded into the main thread, maybe it would be possible to pass the `getIsValidVoiceHandle()` callback to the Web Worker?
Another solution could be to use a thread in the native part where you can monitor the state of the sounds and call a listener in Flutter from C.

**Comment by alnitak** [comment]
> This is probably a stupid question but I want to be clear on something before I ask around. Is there any chance that, on the web, we might not need to use a separate isolate?

There are pros and cons:
- **pros**: it will work
- **cons**: 
  - loading a file¹ will be managed by the main thread making the UI freeze
  - there will be no events to listen to², like `AudioSource.soundEvents` or `AudioSource.allInstancesFinished`

So the only way I see, is to get rid of the Worker on the web, also because the SoLoud lib loaded in the main thread is not accessible from the Web Worker. And loading the lib from the Worker, will not be the same instance of the latter in the main thread.

I may see a **4th** point: make a `render loop` thread in C/C++ that will take the place of `audio_isolate.dart` which will be notified with events (ie load file) or will call Dart if an event occurs (sound disposed). I know this is possible but I never did such a thing. Of course, these will not affect the functionalities exposed to the devs.


¹ on the web is not possible to directly read a file with its path. The devs must use ie FilePicker to choose one (or an assets sound), read the file as bytes and pass them to a newly implemented function `loadMem()`.

² the events are triggered by reading the sound state inside the isolate/worker loop every 10 ms. It wouldn't be feasible to do this in the main thread.

**Comment by filiph** [comment]
Thanks! That's helpful. So I guess it's between:

1. No web support
2. Web support via WebAudio
3. Implement render loop in C/C++ and get web support

Now, for PROs & CONs (advantages & disadvantages) of each approach. Please feel free to edit this.

* **No web support**
  * PROs
    * it's not costing us anything
  * CONs
    * it's going to limit `flutter_soloud`'s use
* Web support via **WebAudio**
  * PROs
    * It's a proven technology. I even have access to someone's private repo where they do exactly this (SoLoud for desktop, WebAudio for web).
  * CONs
    * Harder to add features since all features must be present both in SoLoud and in WebAudio
    * Basically limits us to only a subset of SoLoud. (This is fine by me, though, since this subset is exactly what most game developers will actually need.)
* Implement **render loop in C/C++** and get web support
  * PROs
    * Potentially unlocks all of SoLoud on the web.
  * CONs
    * Potentially lots of work.
    * Unproven approach. We don't know if this will work. 

Does that sound correct? What's your take?

**Comment by alnitak** [comment]
Thanks for this summary, taking a break from coding is useful every now and then to clear my mind.

- **No web support**
Agree with this scenario.

- Web support via **WebAudio**

I think that by using `WebAudio` you mean to write a separate code in JS. Actually when compiling SoLoud lib and miniaudio to WASM, under the hood, the audio backend uses WebAudio. I mean that the example of this repo uses WebAudio.

> It's a proven technology. I even have access to someone's private repo where they do exactly this (SoLoud for desktop, WebAudio for web).

This could be a chance to have audio on the `flutter_soloud` for the web in a short-term way. I leave this decision to you without any problem. But my dream is to have one code base "to rule them all" (cit. :) ) and probably I'll not be able to integrate the WebAudio part of that repo.

- Implement **render loop thread in C/C++** and get web support 

I am exploring this way and as you wrote, we don't know if this will work.
I have seen that the Flutter team added [NativeCallable](https://api.dart.dev/stable/3.2.0/dart-ffi/NativeCallable/NativeCallable.listener.html) which simplifies the communication between separate threads (ie Dart main thread and another thread on C side). But this creates an Isolate on Dart which is not compatible on the web. On the other hand, implementing the render loop on C side, could simplify the `flutter_soloud` for the web (?).


Maybe things could be different when/if an external WASM lib will be supported by acting like does dart:ffi with C.

---
> P.S.: Silly question, but how does the render loop approach interface with long-running processes (such as loading files)? Are we sure this approach won't block / stutter audio when we're loading a file, for example?

The audio has its own thread on all platforms, so even if the main thread is blocked by a long-running process, the audio should still play without stuttering. For example, if you hit a breakpoint on Flutter while playing something, you will still listen to the audio.

Things may be different when loading audio files. Since this means reading its bytes and sending them to SoLoud from the main thread, it will eventually block the UI, but just for a few ms. Maybe this already happens using Web Audio, don't know with certainty.
This could be prevented by loading all game assets sounds at start time.

So my thoughts if we want a relatively fast way to have web support are:
1. use WebAudio from the repo you mentioned.
2. continue with this repo approach giving up on sound events and maybe a little lag when loading files. This will give us all the other `flutter_soloud` functionalities but the experimental `get2DAudioTexture()` I think. Since the Flutter team is moving fast, things may change and we could think of other approaches for these issues.

I'll continue to try to see if there is a common solution within platforms to send messages from C to Dart (ie sound ended or file loaded). Because the root of the problem is exactly this and not really related to audio.

Sorry for the long reply!

**Comment by filiph** [comment]
Thanks for your patience! I was under the weather all last week.

Let's try approach no. 2 (_continue with this repo approach giving up on sound events and maybe a little lag when loading files_), at least for now. I think it's an okay compromise, and we don't need to reimplement anything.

I'm thinking that maybe we should drop `get2DAudioTexture()` and similar experiments. Not that it's _required_ to drop them, per se, but I think it'll make decisions like these easier in the future. People who want to use capture functionality can always fork. But I think they'll be in the minority. If you agree, the next step would be to mark all of `SoLoudCapture` as `@Deprecated`.
